src/ui/ui_settings/ui_settings_mixin.js

```diff
@@ -9,9 +10,35 @@ export function uiSettingsMixin(kbnServer, server, config) {
     return;
   }
 
-  const uiSettings = new UiSettings(server, status);
-  server.decorate('server', 'uiSettings', () => uiSettings);
+  // Passed to the UiSettingsService.
+  // UiSettingsService calls the function before trying to read data from
+  // elasticsearch, giving us a chance to prevent it from happening.
+  //
+  // If the ui settings status isn't green we shouldn't be attempting to get
+  // user settings, since we can't be sure that all the necessary conditions
+  // (e.g. elasticsearch being available) are met.
+  const readUiSettingsInterceptor = () => {
+    if (status.state !== 'green') {
+      return {};
+    }
+  };
+
+  // don't return, just let it happen when the plugins are ready
   kbnServer.ready().then(() => {
     mirrorStatus(status, kbnServer.status.getForPluginId('elasticsearch'));
   });
+
+  server.decorate('server', 'uiSettingsServiceFactory', function (options) {
+    return uiSettingsServiceFactory(server, options);
+  });
+
+  server.decorate('request', 'getUiSettingsService', function () {
+    return getUiSettingsServiceForRequest(server, this, readUiSettingsInterceptor);
+  });
+
+  server.decorate('server', 'uiSettings', () => {
```

Comment:
__question:__ if something like this is still used in x-pack, how do you usually manage it? Prepare follow-up PR as soon as this one is merged or notify owners of that x-pack code about the change so that they can update code?